alpha/template/template.go

```diff
@@ -0,0 +1,115 @@
+package template
+
+import (
+	"context"
+	"io"
+	"slices"
+	"strings"
+
+	"github.com/operator-framework/operator-registry/alpha/declcfg"
+)
+
+// BundleRenderer defines the function signature for rendering bundle images
+type BundleRenderer func(context.Context, string) (*declcfg.DeclarativeConfig, error)
+
+// Template defines the common interface for all template types
+type Template interface {
+	// RenderBundle renders a bundle image reference into a DeclarativeConfig
+	RenderBundle(ctx context.Context, image string) (*declcfg.DeclarativeConfig, error)
+	// Render processes the template input and returns a DeclarativeConfig
+	Render(ctx context.Context, reader io.Reader) (*declcfg.DeclarativeConfig, error)
+	// Schema returns the schema identifier for this template type
+	Schema() string
+}
+
+// TemplateFactory creates template instances based on schema
+type TemplateFactory interface {
+	// CreateTemplate creates a new template instance with the given RenderBundle function
+	CreateTemplate(renderBundle BundleRenderer) Template
+	// Schema returns the schema identifier this factory handles
+	Schema() string
+}
+
+// TemplateRegistry maintains a mapping of schema identifiers to template factories
+type TemplateRegistry struct {
+	factories map[string]TemplateFactory
+}
+
+// NewTemplateRegistry creates a new template registry
+func NewTemplateRegistry() *TemplateRegistry {
+	return &TemplateRegistry{
+		factories: make(map[string]TemplateFactory),
+	}
+}
+
+// Register adds a template factory to the registry
+func (r *TemplateRegistry) Register(factory TemplateFactory) {
+	r.factories[factory.Schema()] = factory
+}
+
+// CreateTemplateBySchema creates a template instance based on the schema found in the input
+// and returns a reader that can be used to render the template. The returned reader includes
+// both the data consumed during schema detection and the remaining unconsumed data.
```

Comment:
could returned reader be hidden inside the returned Template instance? can this reader be used independently?

Reply:
It could be hidden in the template type, but it's arbitrary. The template is only interested in the reader as an input source, but it isn't inherently a template concern.  It's just a general I/O one.